net_processing: lock clean up

ajtowns opened the pull request:
`cs_sendProcessing` is replaced by a private mutex in net_processing, non-orphan-specific things are moved out from `g_cs_orphans` and `g_cs_orphans` is replaced by a private mutex in txorphanage.

ajtowns committed:
net_processing: Pass a Peer to ProcessOrphanTx

ajtowns committed:
net_processing: Localise orphan_work_set handling to ProcessOrphanTx

ajtowns committed:
txorphange/net_processing: move orphan workset to txorphanage

ajtowns committed:
txorphanage: make m_peer_work_set private

ajtowns committed:
Move all g_cs_orphans locking to txorphanage

ajtowns committed:
txorphanage: move g_cs_orphans to TxOrphanage::m_mutex

jnewbery commented on src/txorphanage.h:
```diff
@@ -21,65 +23,75 @@ extern RecursiveMutex g_cs_orphans;
+     *  -maxorphantx/DEFAULT_MAX_ORPHAN_TRANSACTIONS */
+    OrphanMap m_orphans GUARDED_BY(m_mutex);
+
+    /** Which peer provided a parent tx of orphans that need to be reconsidered */
```
This comment is now incorrect. The multimap stores the node that provided the orphan tx.

Also, this data is redundant data with the `fromPeer` in the `OrphanTx` struct (this could easily be a set of txids to reconsider, and the originating peer is then looked up in the `OrphanTx` struct). We store it in the multimap for efficiency - the comment could be updated to indicate that.

jnewbery reviewed (commented):
Concept ACK. Lots of minor review comments inline, but a bit of high-level design feedback:

- I'm not at all convinced by the `g_mutex_message_handler`. It seems like you're trying to reinvent thread_local storage. That makes potential future changes difficult. It also adds unnecessary coupling between net and net_processing.
- This PR changes orphan processing from being done for the peer that provided the parent to the peer that provided the orphan. I think that's fine, but perhaps we should split the PR into one that does refactoring only and one that has behaviour changes.

ajtowns commented on src/net_processing.cpp:
```diff
@@ -3059,7 +3060,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
             // Recursively process any orphan transactions that depended on this one
-            ProcessOrphanTx(peer->m_orphan_work_set);
+            ProcessOrphanTx(*peer);
```
We could process 3 txs I think -- the last orphan already in the workset (in ProcessMessages), one from a just received TX message, and an additional orphan whose parent was the contents of that TX message. I don't think it's problematic that way, but 1-non-trivial-ATMP-call per ProcessMessages invocation could be reasonable.

jnewbery replied:
Very good point. I hadn't considered processing the orphan _before_ processing the net message.

ajtowns commented on src/txorphanage.cpp:
```diff
@@ -87,24 +86,27 @@ int TxOrphanage::EraseTx(const uint256& txid)
-    AssertLockHeld(g_cs_orphans);
+    LOCK(m_mutex);
+
+    const int work_erased = m_peer_work_set.erase(peer);
```
It crossed my mind that this might be another case where a multi index could be a win.

ajtowns commented on src/net.h:
```diff
@@ -232,6 +232,7 @@ struct LocalServiceInfo {
     uint16_t nPort;
 };
 
+extern Mutex g_mutex_message_handler;
```
The reasoning isn't that this is particularly "good", it's that it's documenting the safety mechanisms we already have. It would have caught https://github.com/bitcoin/bitcoin/pull/21528#discussion_r601317221 for instance.

net_processing is fundamentally coupled with net -- it implements the NetEventsInterface defined and called from net. I mean I guess if you preferred, it could be `NetEventsInterface::m_mutex_message_handler`?

jnewbery replied:
And if you're doing that then perhaps you can make it a `PeerManagerImpl::m_mutex_message_handler` and take it in `ProcessMessages()` and `SendMessages()`. Taking that lock once at the top of those functions would be very low cost (since there would never be any contention).

ajtowns replied:
I've done this now -- the first three commits introduce `m_mutex_messages`, remove `cs_sendProcessing` and move extra txns to be protected by `m_mutex_messages` instead of `g_cs_orphans`. I think it would be good to cherry-pick those commits into #21186 so that the moved addr variables can actually be marked as guarded by something.

jnewbery replied:
Also see #21563

ajtowns commented on src/net_processing.cpp:
```diff
@@ -2010,33 +2016,25 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
-        const auto [porphanTx, from_peer] = m_orphanage.GetTx(orphanHash);
-        if (porphanTx == nullptr) continue;
+    while (m_orphanage.GetTxToReconsider(peer.m_id, porphanTx, from_peer, more)) {
+        if (porphanTx == nullptr) break;
```
`assert(porphanTx != nullptr)` seemed overkill and not having anything seemed like it might be annoying for static analysers

jnewbery replied:
How about: `if (!Assume(porphanTx)) break` ?

ajtowns commented on src/init.cpp:
```diff
@@ -203,15 +203,14 @@ void Shutdown(NodeContext& node)
     if (node.connman) {
         node.connman->StopThreads();
-        LOCK2(::cs_main, ::g_cs_orphans);
+        LOCK(::cs_main);
```
I updated the comments which should still indicate why cs_main is needed there -- it needs to be locked before cs_vNodes, but cs_vNodes is locked first if you call StopThreads directly?

jnewbery replied:
Ah, very good point. And we can't take `cs_main` inside `StopNodes` because net isn't aware of `cs_main`.

It's outside the scope of this PR, but it'd be good to untangle this lock ordering dependency by eg making sure that cs_vNodes is not held when calling any PeerManager functions.

jnewbery replied:
See #21563

jnewbery replied:
21563 is merged

ajtowns committed:
net: add NetEventsInterface::g_mutex_msgproc_thread mutex

There are many cases where we asssume message processing is
single-threaded in order for how we access node-related memory to be
safe. Add an explicit mutex that we can use to document this, which allows
the compiler to catch any cases where we try to access that memory from
other threads and break that assumption.

ajtowns committed:
net_processing: extra transactions data are accessed only via the msgproc thread

Previously vExtraTxnForCompact and vExtraTxnForCompactIt were protected
by g_cs_orphans; protect them by g_mutex_msgproc_thread instead, as they
are only used during message processing.

ajtowns committed:
net: drop cs_sendProcessing

SendMessages() is now protected g_mutex_msgproc_thread; so this additional
per-node mutex is redundant.

ajtowns commented on src/net_processing.cpp:
```diff
@@ -2010,33 +2016,25 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
-void PeerManagerImpl::ProcessOrphanTx(std::set<uint256>& orphan_work_set)
+bool PeerManagerImpl::ProcessOrphanTx(Peer& peer)
 {
     AssertLockHeld(cs_main);
```
I think:

 * just call AddChildrenToWorkSet when a new tx arrives, don't also immediately process them (ie don't also call ProcessOrphanTx)
 * when ProcessOrphanTx does work, don't continue on to also do a ProcessMessage, give another peer a chance to get some work done first
 * have ProcessOrphanTx take cs_main only when necessary
 * make the GetTxToReconsider signature simpler
 
would all fit together nicely and be a good idea, but it also feels like it's adding a bit much to this PR?

ajtowns replied:
> This PR changes orphan processing from being done for the peer that provided the parent to the peer that provided the orphan. I think that's fine, but perhaps we should split the PR into one that does refactoring only and one that has behaviour changes.

Hmm, might make sense to put the changes above together with the last commit in a separate PR, so this is just refactoring and the behaviour changes (who processes orphans, and how many orphans are processed in a cycle) are collected together.

jnewbery replied:
Yes, I agree that it's better to split the PR into two: one that refactors and one that changes behaviour.

ajtowns replied:
This is drafted at https://github.com/ajtowns/bitcoin/commits/202104-whohandlesorphans

ajtowns commented:
Rebased, some of the suggestions picked up, reworked the "message handler" mutex.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -253,9 +250,18 @@ class PeerManagerImpl final : public PeerManager
+     */
+    Mutex m_mutex_messages;
+
+    void _ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRecv,
```
I think it'd be cleaner for the `m_mutex_messages` lock to be taken by ProcessMessage**s**, rather than ProcessMessage. That would remove the need to define an inner `_ProcessMessage` to be called when the mutex is already locked. It's also consistent with the comment on the `m_mutex_messages` member, which says "is acquired by ProcessMessages and SendMessages"

ajtowns replied:
It is taken by `ProcessMessages()` (and also by `SendMessages()`), however `ProcessMessage()` is also called from the test suite, so cannot assume the lock has been taken. That's why the public interface, `ProcessMessage()`, is implemented as a wrapper that takes the lock, then calls the internal function.

jnewbery replied:
Ah, sorry - misread.

jnewbery reviewed (commented):
I don't understand the second commit log: _SendMessages() is now protected internally by m_mutex_messages; so this additional lock is not useful._ `SendMessages()` is not protected by the new mutex as far as I can see.

I prefer the approach in #21563, which locks the mutex whenever any `NetEvents` method is called.

jnewbery commented:
I've re-reviewed the locking changes and they look reasonable. There are still a few review comments outstanding. I'm happy to review this PR again once those have been addressed.

ajtowns commented:
I think this is now cleaned up, and ready for review. The behaviour changes that were in the last commit are now deferred to https://github.com/ajtowns/bitcoin/commits/202104-whohandlesorphans .

jnewbery commented:
When you rebase this, there are a few files that no longer need to include txorphanage.h:

<details>
<summary>Diff</summary>

```diff
diff --git a/src/init.cpp b/src/init.cpp
index bb5b144802..e7b5ed60e3 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -52,7 +52,6 @@
 #include <torcontrol.h>
 #include <txdb.h>
 #include <txmempool.h>
-#include <txorphanage.h>
 #include <util/asmap.h>
 #include <util/check.h>
 #include <util/moneystr.h>
diff --git a/src/test/fuzz/process_message.cpp b/src/test/fuzz/process_message.cpp
index 7b99193ad0..5a71b25768 100644
--- a/src/test/fuzz/process_message.cpp
+++ b/src/test/fuzz/process_message.cpp
@@ -18,7 +18,6 @@
 #include <test/util/net.h>
 #include <test/util/setup_common.h>
 #include <test/util/validation.h>
-#include <txorphanage.h>
 #include <validationinterface.h>
 #include <version.h>
 
diff --git a/src/test/fuzz/process_messages.cpp b/src/test/fuzz/process_messages.cpp
index 11b236c9bd..f8b1c8fc90 100644
--- a/src/test/fuzz/process_messages.cpp
+++ b/src/test/fuzz/process_messages.cpp
@@ -13,7 +13,6 @@
 #include <test/util/net.h>
 #include <test/util/setup_common.h>
 #include <test/util/validation.h>
-#include <txorphanage.h>
 #include <validation.h>
 #include <validationinterface.h>
```

</details>

Those files were only including txorphange.h to use the g_cs_orphans mutex.

jnewbery committed:
Remove unnecessary includes of txorphanage.h

ajtowns commented:
Rebased (and likewise the whohandlesorphans followup)

jnewbery commented on src/txorphanage.h:
```diff
@@ -21,65 +21,75 @@ extern RecursiveMutex g_cs_orphans;
 
-    /** Get an orphan transaction and its originating peer
-     * (Transaction ref will be nullptr if not found)
+    /** Get an orphan transaction for a peer to work on
```
```suggestion
    /** Get an orphan transaction for a peer to work on and erase it from the peer's workset.
```

ajtowns replied:
Done-ish

jnewbery replied:
thank you!

jnewbery reviewed (commented):
utACK 584043a0203332fe645529b1c27c086e4f14a61c

Very nice cleanup.

ajtowns commented:
Rebased past #21845. cc @MarcoFalke @promag for reviews :)

jnewbery commented:
utACK 8b4f685ebef5eb14f049d04e2c4f4ce5b44878de

Do you mind addressing https://github.com/bitcoin/bitcoin/pull/21527#discussion_r622998740 while there are no other ACKs on the branch?

jnewbery commented:
ACK dd754fd73ab0fcbe9461a6a21b5fc2bc22faf968

MarcoFalke commented on src/txorphanage.h:
```diff
@@ -43,9 +43,11 @@ class TxOrphanage {
-     * (ie orphans that may have found their final missing parent, and so should be reconsidered for the mempool) */
-    void AddChildrenToWorkSet(const CTransaction& tx, std::set<uint256>& orphan_work_set) const EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
+    /** Which peer provided a parent tx of orphans that need to be reconsidered */
+    std::multimap<NodeId, uint256> m_peer_work_set GUARDED_BY(g_cs_orphans);
```
67618f024690bc9b926aa48ead23a0f687f03fe4: I know there is a check to avoid duplicate insertion of the same txid, but is there a reason to pick this datastructure, which doesn't disallow duplicate entries like set? Also insertion is trivially more expensive (log(all peers' work sets) vs log(this peers' work set)). Finally handling is done in insertion order, not in sorted order.

vasild replied:
> ... is there a reason to pick this datastructure, which doesn't disallow duplicate entries ...

I think the reason could be to allow (nodeX, tx1), (nodeX, tx2)?

ajtowns replied:
Note that if you've got N peers each with K items in their work set, then a log search over everything is `log(N*K)`, and a log search for the peer followed by a log search or the work item is `log(N) + log(K)` which is the same. If there's one peer with K items and every other peer has 0, you instead get `log(K)` vs `log(N) + 1/N log(K)`; so a map of sets would be better than a set of pairs provided `N < K`, but probably isn't very interesting.

I picked the data structure because I thought it was simpler than a set of pairs, but I don't think there's an adequate check to avoid duplicate insertion of the same txid (and to do that, you'd have to search through all the entries for a peer, which would be annoying).

ajtowns replied:
Changed to `map<NodeId,set<uint256>>`

maflcko reviewed (commented):
dd754fd73ab0fcbe9461a6a21b5fc2bc22faf968 🕵

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

dd754fd73ab0fcbe9461a6a21b5fc2bc22faf968 🕵
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgcDAwAqL240wertSr+Mw2Ei6s6SQizRd2AzRniN6bec0MgUdiNK8aOz7CBTuf5
3g5Jt5vqORpMkWidS6+JZFMDJQzOzcZXet+eGP6OpuUEFjUKQ2DXIPvRUvtZZU0s
d0Ad75v5Ov/jw5dU/+Mh1o/eaNJvtvE0bkHJMZrounpUDIH+0M0XKuzCVDPe1NeN
PomxQwCDlCBbpYJ02ca6oOo9mSzVi0IaMToCdb+S0nI3/1fZwJ2saN16w1ul9Xg+
8w8Z6j9sxvV9DTTkQTVqv7ckyaXO7V/pZVGzW2FI2hqHYSf8LLRae07+Muqxlec7
6ldtgIZE0v262nVbcWJFITGr6hrZPr/Oyo6FBfAZnNjZmr/7Z60hY2n2FSNu8GUG
B84+424mEkH2oOcevu6GeP/GKoL/JXu1gLNxKKPUoOohBfTMHM1OfAeEVm/o/DnI
xIhl10GoHg3jo127zutHb1TCvknUKCzJzPeHbVvUSJB9g5DIbKqPaABFn7k+cCUy
zJ3COLUx
=f8yk
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `04c8a7094b10796be32033da7a78c70f9926aef2456fc44aaece53aa2519e97d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010804c8a7094b10796be32033da7a78c70f9926aef2456fc44aaece53aa2519e97df010e6984a50046b85f9751ce02adcc0381f08fff0108899f33747c05cd789c99839c0fc2acb08f020328a9c818e7ca87409f85bc29cadf8934eb564d2567d4db2cd171de23a797ad808f10460977b8df00884974b7dda4baa310083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01098ff5aff5cbbe59ecd8b2f63f2f7383a08f10460977b8df008da426e455d56a9790083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010bd02bb5dddaa70666f9c822eb55c30ec08f120e675d46d23a2e425ee65d2ac5a634712685cb1fc6ed97e3457f178223ebfcd3408f10460977b8df008f8fd6ab1fd94c1570083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

ajtowns commented on src/net_processing.cpp:
```diff
@@ -3895,9 +3902,7 @@ bool PeerManagerImpl::ProcessMessages(CNode* pfrom, std::atomic<bool>& interrupt
-        if (!peer->m_orphan_work_set.empty()) {
-            ProcessOrphanTx(*peer);
-        }
+        if (ProcessOrphanTx(*peer)) return true;
```
Not sure what you mean by "the comment is gone" ? The "maintains the order of responses and prevents vRecvGetData to grow unbounded" is still there, but at least I thought that sensibly described the `!pfrom->vRecvGetData.empty()` check it remains attached to, and didn't really fit the "more orphans to work on" check.

ajtowns replied:
Added a comment.

vasild commented on src/net_processing.cpp:
```diff
@@ -254,9 +251,20 @@ class PeerManagerImpl final : public PeerManager
+                        const std::chrono::microseconds time_received, const std::atomic<bool>& interruptMsgProc) override LOCKS_EXCLUDED(m_mutex_message_handling);
 
 private:
+    /** Message handling mutex.
+     *  Message processing is single-threaded, so anything only accessed
+     *  by ProcessMessage(s) or SendMessages can be guarded by this mutex,
+     *  which guarantees it's only accessed by a single thread.
+     */
+    Mutex m_mutex_message_handling;
```
If it is single-threaded, then we don't need to guard anything with a mutex?

It looks to me that `PeerManagerImpl::m_mutex_message_handling` is useless because it is only being acquired from a single thread. No two threads are ever going to race to acquire it at the same time:

```cpp
ThreadMessageHandler() // there is just one thread executing this
   ProcessMessages()
       ...
       lock m_mutex_message_handling
       ...
   ...
   SendMessages()
       ...
       lock m_mutex_message_handling
       ...
```

I may be missing something but to me it looks like `m_mutex_message_handling` can be removed safely, which will reduce the complexity of the code.

ajtowns replied:
You don't need the mutex at runtime, but having the `GUARDED_BY(m_mutex_message_handling)` at compile time makes it easy to verify nobody's accidentally accessing the object from some other thread (either currently, or as a result of some future change, eg adding an RPC call that wants to expose some of the information stored in those objects). The cost of also having the mutex at runtime is trivial since there's never any contention on it.

vasild replied:
NACK the addition of `m_mutex_message_handling`.

Following the logic of

> GUARDED_BY() makes it easy to verify nobody's accidentally accessing...

should we add a dummy mutex for **every** variable?

IMO mutexes should be used only where concurrent access is possible.

jnewbery replied:
@vasild can you take a look at the commits that build on top of this branch in https://github.com/jnewbery/bitcoin/commits/2021-05-use-mutex-message-handling. Currently, a lot of the data in net_processing is guarded by cs_main, which is also used in validation. One way to loosen that coupling between validation and net_processing is to introduce an internal mutex inside net_processing (`m_mutex_message_handling`) that guards its own members.

vasild replied:
I think that https://github.com/jnewbery/bitcoin/commits/2021-05-use-mutex-message-handling can be simplified and the common `m_mutex_message_handling` avoided by changing some of the variables to atomics and using a dedicated mutex for e.g. `m_txrequest`.

Btw, it does not compile.
<details>
<summary>minor fixups</summary>

```diff
diff --git i/src/net_processing.cpp w/src/net_processing.cpp
index c56863e5ef..cc04679683 100644
--- i/src/net_processing.cpp
+++ w/src/net_processing.cpp
@@ -401,13 +401,13 @@ private:
      * Set mapBlockSource[hash].second to false if the node should not be
      * punished if the block is invalid.
      */
     std::map<uint256, std::pair<NodeId, bool>> mapBlockSource GUARDED_BY(cs_main);
 
     /** Number of peers with wtxid relay. */
-    int m_wtxid_relay_peers GUARDED_BY(m_mutex_message_handing) {0};
+    int m_wtxid_relay_peers GUARDED_BY(m_mutex_message_handling) {0};
 
     /** Number of outbound peers with m_chain_sync.m_protect. */
     int m_outbound_peers_with_protect_from_disconnect GUARDED_BY(cs_main) = 0;
 
     bool AlreadyHaveTx(const GenTxid& gtxid) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
@@ -1000,20 +1000,15 @@ void UpdateLastBlockAnnounceTime(NodeId node, int64_t time_in_seconds)
     CNodeState *state = State(node);
     if (state) state->m_last_block_announcement = time_in_seconds;
 }
 
 void PeerManagerImpl::InitializeNode(CNode *pnode)
 {
-    LOCK(m_mutex_message_handling);
-
     NodeId nodeid = pnode->GetId();
-    {
-        LOCK(cs_main);
-        mapNodeState.emplace_hint(mapNodeState.end(), std::piecewise_construct, std::forward_as_tuple(nodeid), std::forward_as_tuple(pnode->IsInboundConn()));
-        assert(m_txrequest.Count(nodeid) == 0);
-    }
+    WITH_LOCK(cs_main, mapNodeState.emplace_hint(mapNodeState.end(), std::piecewise_construct, std::forward_as_tuple(nodeid), std::forward_as_tuple(pnode->IsInboundConn())));
+    WITH_LOCK(m_mutex_message_handling, assert(m_txrequest.Count(nodeid) == 0));
     {
         PeerRef peer = std::make_shared<Peer>(nodeid, m_mutex_message_handling);
         LOCK(m_peer_mutex);
         m_peer_map.emplace_hint(m_peer_map.end(), nodeid, std::move(peer));
     }
     if (!pnode->IsInboundConn()) {
```
</details>

<details>
<summary>compilation errors</summary>

```sh
(clang 13)

net_processing.cpp:986:33: error: reading variable 'fPreferredDownload' requires holding mutex 'peer.m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
    const bool preferred = peer.fPreferredDownload;
                                ^
net_processing.cpp:1033:13: error: calling function '_RelayTransaction' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
            _RelayTransaction(txid, tx->GetWitnessHash());
            ^
net_processing.cpp:1062:41: error: reading variable 'fPreferredDownload' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
            nPreferredDownload -= peer->fPreferredDownload;
                                        ^
net_processing.cpp:1063:42: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
            m_wtxid_relay_peers -= peer->m_wtxid_relay;
                                         ^
net_processing.cpp:1537:41: error: calling function '_RelayTransaction' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
    WITH_LOCK(m_mutex_message_handling, _RelayTransaction(txid, wtxid););
                                        ^
net_processing.cpp:1550:43: error: reading variable 'm_wtxid_relay' requires holding mutex 'it->second->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
        node->PushTxInventory(it->second->m_wtxid_relay ? wtxid : txid);
                                          ^
net_processing.cpp:2485:15: error: writing variable 'fPreferredDownload' requires holding mutex 'peer->m_mutex_message_handling' exclusively [-Werror,-Wthread-safety-analysis]
        peer->fPreferredDownload = (!pfrom.IsInboundConn() || pfrom.HasPermission(PF_NOBAN)) && !pfrom.IsAddrFetchConn() && !pfrom.fClient;
              ^
net_processing.cpp:2486:37: error: reading variable 'fPreferredDownload' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
        nPreferredDownload -= peer->fPreferredDownload;
                                    ^
net_processing.cpp:2653:24: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
            if (!peer->m_wtxid_relay) {
                       ^
net_processing.cpp:2654:23: error: writing variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' exclusively [-Werror,-Wthread-safety-analysis]
                peer->m_wtxid_relay = true;
                      ^
net_processing.cpp:2777:23: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
            if (peer->m_wtxid_relay) {
                      ^
net_processing.cpp:3049:37: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
        const uint256& hash = peer->m_wtxid_relay ? wtxid : txid;
                                    ^
net_processing.cpp:3051:19: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
        if (peer->m_wtxid_relay && txid != wtxid) {
                  ^
net_processing.cpp:3075:13: error: calling function 'AlreadyHaveTx' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
        if (AlreadyHaveTx(GenTxid(/* is_wtxid=*/true, wtxid))) {
            ^
net_processing.cpp:3084:21: error: calling function '_RelayTransaction' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
                    _RelayTransaction(tx.GetHash(), tx.GetWitnessHash());
                    ^
net_processing.cpp:3090:44: error: calling function 'AcceptToMemoryPool' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
        const MempoolAcceptResult result = AcceptToMemoryPool(m_chainman.ActiveChainstate(), m_mempool, ptx, false /* bypass_limits */);
                                           ^
net_processing.cpp:3094:23: error: calling function 'check' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
            m_mempool.check(m_chainman.ActiveChainstate());
                      ^
net_processing.cpp:3099:13: error: calling function '_RelayTransaction' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
            _RelayTransaction(tx.GetHash(), tx.GetWitnessHash());
            ^
net_processing.cpp:3114:13: error: calling function 'ProcessOrphanTx' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
            ProcessOrphanTx(*peer);
            ^
net_processing.cpp:3131:21: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                if (recentRejects->contains(parent_txid)) {
                    ^
net_processing.cpp:3147:26: error: calling function 'AlreadyHaveTx' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
                    if (!AlreadyHaveTx(gtxid)) AddTxAnnouncement(pfrom, *peer, gtxid, current_time);
                         ^
net_processing.cpp:3172:17: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                recentRejects->insert(tx.GetHash());
                ^
net_processing.cpp:3173:17: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                recentRejects->insert(tx.GetWitnessHash());
                ^
net_processing.cpp:3192:24: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                assert(recentRejects);
                       ^
net_processing.cpp:3193:17: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                recentRejects->insert(tx.GetWitnessHash());
                ^
net_processing.cpp:3204:21: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                    recentRejects->insert(tx.GetHash());
                    ^
net_processing.cpp:4291:29: error: reading variable 'fPreferredDownload' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
        bool fFetch = peer->fPreferredDownload || (nPreferredDownload == 0 && !pto->fClient && !pto->IsAddrFetchConn()); // Download if this is a nice peer, or we have no nice peers and this one might do.
                            ^
net_processing.cpp:4504:53: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
                        const uint256& hash = peer->m_wtxid_relay ? txinfo.tx->GetWitnessHash() : txinfo.tx->GetHash();
                                                    ^
net_processing.cpp:4505:40: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
                        CInv inv(peer->m_wtxid_relay ? MSG_WTX : MSG_TX, hash);
                                       ^
net_processing.cpp:4536:85: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
                    CompareInvMempoolOrder compareInvMempoolOrder(&m_mempool, peer->m_wtxid_relay);
                                                                                    ^
net_processing.cpp:4548:40: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
                        CInv inv(peer->m_wtxid_relay ? MSG_WTX : MSG_TX, hash);
                                       ^
net_processing.cpp:4636:117: error: reading variable 'fPreferredDownload' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
                if (current_time > state.m_headers_sync_timeout && nSyncStarted == 1 && (nPreferredDownload - peer->fPreferredDownload >= 1)) {
                                                                                                                    ^
32 errors generated.
```
</details>

ajtowns replied:
"Should we add a dummy mutex for every variable" -- we should have a GUARDED_BY for every object that's reachable by multiple threads, yes. The mutex is not a dummy, it's a real mutex.

For this specific case, here's an example of a patch that adds an access from another thread: https://github.com/ajtowns/bitcoin/commits/202105-whyhaveaguard . With the `GUARDED_BY(m_mutex_message_handling)` this is caught at compile time; without the guard (removed in the second commit), it compiles fine but introduces a race condition.

vasild replied:
> we should have a GUARDED_BY for every object that's reachable by multiple threads

I assume here you mean -- even if the current code accesses it from a single thread.

If yes, then I think that is an overkill - it means having a mutex attached to every non-const global or class member variable, including private ones (coz in the future somebody may access it from another thread).

> For this specific case, here's an example...

Yeah, but one can do such an example for any other variable. IMO annotating everything with `GUARDED_BY()`, just in case, would have a net-adverse effect on the code base.

<details>
<summary>Similar example for two other, randomly picked variables</summary>

```diff
diff --git i/src/net.h w/src/net.h
index 6b9a7ef136..18d57a7d8e 100644
--- i/src/net.h
+++ w/src/net.h
@@ -1256,12 +1256,18 @@ private:
      * an address and port that are designated for incoming Tor connections.
      */
     std::vector<CService> m_onion_binds;
 
     friend struct CConnmanTest;
     friend struct ConnmanTestMsg;
+
+public:
+    size_t ResponseCachesSize() const
+    {
+        return m_addr_response_caches.size();
+    }
 };
 
 /** Return a timestamp in the future (in microseconds) for exponentially distributed events. */
 std::chrono::microseconds PoissonNextSend(std::chrono::microseconds now, std::chrono::seconds average_interval);
 
 /** Dump binary message to file, with timestamp */
diff --git i/src/rpc/net.cpp w/src/rpc/net.cpp
index 1f6b6e8d7e..cfab1ebafc 100644
--- i/src/rpc/net.cpp
+++ w/src/rpc/net.cpp
@@ -631,12 +631,14 @@ static RPCHelpMan getnetworkinfo()
     NodeContext& node = EnsureAnyNodeContext(request.context);
     if (node.connman) {
         ServiceFlags services = node.connman->GetLocalServices();
         obj.pushKV("localservices", strprintf("%016x", services));
         obj.pushKV("localservicesnames", GetServicesNames(services));
     }
+    obj.pushKV("bug1", node.connman->ResponseCachesSize());
+    obj.pushKV("bug2", node.addrman->m_asmap.size());
     if (node.peerman) {
         obj.pushKV("localrelay", !node.peerman->IgnoresIncomingTxs());
     }
     obj.pushKV("timeoffset",    GetTimeOffset());
     if (node.connman) {
         obj.pushKV("networkactive", node.connman->GetNetworkActive());
```
</details>

ajtowns replied:
> `diff --git i/src/net.h w/src/net.h`

Yes, net.cpp/net.h is terrible for having globally accessible variables that aren't annotated for thread safety. Being able to introduce races that the compiler doesn't catch is a bad thing... txmempool's a better example, though it's also missing some guards (cf #22003).

> IMO annotating everything with `GUARDED_BY()`, just in case, would have a net-adverse effect on the code base.

`GUARDED_BY` is always a "just in case" thing -- any time the code compiles correctly with `GUARDED_BY` it will also compile correctly without it -- the benefit is "just in case" someone forgets while modifying the code later. With a guard, the compiler will catch the mistake; without it, you have to hope someone reviewing the code remembers, or wait for some obscure bug to be noticed and reported.

Not consistently guarding variables has an easily observable adverse affect on the code bas as a whole: it makes it hard to change code, because you're not sure when things might be safe to reuse in a different context. eg see https://github.com/bitcoin/bitcoin/pull/21061#discussion_r602283867

>  it means having a mutex attached to every non-const global or class member variable, including private ones

Yes and no -- no, in that most classes don't need to manually manage locks, but should rather rely on whatever instantiates them to ensure the entire object is accessed safely; but for the ones that do, yes: every non-const, non-atomic member should be guarded (as should every class's non-const, non-atomic static members, since they're effectively globals themselves).

vasild replied:
@ajtowns, thanks for the elaboration.

I still think that mutexes should be introduced when needed, not beforehand (with a reason that in the future somebody may access a variable from a different thread).

Looks like neither one of us is convinced by the other's arguments. It's ok, it would have been too boring if everybody agreed on everything all the time.

MarcoFalke replied:
I think it is fine to not have thread safety stuff on things that aren't meant to be called by more than one thread. For example, most code in init.cpp doesn't need them because they are only called once per process. Though, other stuff should communicate to the developer whether it is thread safe. Either via code comments (https://github.com/bitcoin/bitcoin/blob/e3de7cb9039770e0fd5b8bb8a5cba35c87ae8f00/src/random.h#L67, https://github.com/bitcoin/bitcoin/blob/e3de7cb9039770e0fd5b8bb8a5cba35c87ae8f00/src/random.h#L129) or via annotations.

jnewbery commented on src/txorphanage.cpp:
```diff
@@ -132,15 +134,19 @@ unsigned int TxOrphanage::LimitOrphans(unsigned int max_orphans)
-    AssertLockHeld(g_cs_orphans);
+    LOCK(m_mutex);
+
+    // lookup this peer's work set, ensuring it exists
```
```suggestion
    // Look up this peer's work set, ensuring it exists.
```

jnewbery commented on src/txorphanage.h:
```diff
@@ -21,65 +21,76 @@ extern RecursiveMutex g_cs_orphans;
+     *  after this tx.
      */
-    std::pair<CTransactionRef, NodeId> GetTx(const uint256& txid) const EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
+    bool GetTxToReconsider(NodeId peer, CTransactionRef& ref, NodeId& originator, bool& more) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
```
I left a comment about this interface further up that seems to have been lost without being resolved. I think the interface is needlessly complex and doing too much. From my comment above:

> How about returning a `std::optional<std::pair<CTransactionRef, NodeId>>`, and then adding another public method `HasTxToReconsider()`. It's a bit less efficient, but we'll only ever call it after processing a transaction, so it seems negligible.

ajtowns replied:
See https://github.com/bitcoin/bitcoin/pull/21527#discussion_r601221165

jnewbery commented on src/txorphanage.cpp:
```diff
@@ -153,26 +159,43 @@ void TxOrphanage::AddChildrenToWorkSet(const CTransaction& tx, std::set<uint256>
+
+            const auto orphan_it = m_orphans.find(txid);
+            if (orphan_it != m_orphans.end()) {
+                more = (it != work_set.end());
```
(if you're keeping `more`), the following is equivalent, and maybe more direct/clear:

```suggestion
                more = !(work_set.empty());
```

jnewbery reviewed (commented):
utACK d2f5cc9de12196c870b2079bffc38e90076074d4.

Verified range-diff. Only significant difference was replacing `std::multimap<NodeId, uint256> m_peer_work_set` with `std::map<NodeId, std::set<uint256>> m_peer_work_set`.

ajtowns commented:
Rebased past #21186, addressed @jnewbery's nits.

jnewbery commented:
Code review ACK d6dfa5977a7c03d9a81727e1692edb58bfeab62c

vasild commented on src/txorphanage.h:
```diff
@@ -21,65 +21,76 @@ extern RecursiveMutex g_cs_orphans;
 
+    /** Map from txid to orphan transaction record. Limited by
+     *  -maxorphantx/DEFAULT_MAX_ORPHAN_TRANSACTIONS */
+    OrphanMap m_orphans GUARDED_BY(m_mutex);
```
Is it possible for `TxOrphanage::m_orphans` to contain keys (transaction ids) which are not in `TxOrphanage::m_peer_work_set` or the other way around?

ajtowns replied:
`m_orphans` will normally contain txids that aren't in `m_peer_work_set` -- they only get added to the work set when a parent appears and are (hopefully) very quickly removed from the work set after retrying ATMP.

It's also possible for something to expire from `m_orphans` and still be present in `m_peer_work_set`: if a tx that's in the work set is removed via `LimitOrphans` or `EraseForBlock` you'll get that case. It will eventually be removed from the work set when `GetTxToReconsider` is called.

ajtowns commented on src/net_processing.cpp:
```diff
@@ -2132,33 +2152,26 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
 
-        const auto [porphanTx, from_peer] = m_orphanage.GetTx(orphanHash);
-        if (porphanTx == nullptr) continue;
+    while (m_orphanage.GetTxToReconsider(peer.m_id, porphanTx, from_peer, more)) {
+        if (!Assume(porphanTx)) break;
```
All Assume's should be unnecessary by definition? See https://github.com/bitcoin/bitcoin/pull/21527#discussion_r601253387 for prior discussion.

vasild replied:
`Assume()` makes sense if the value is derived in an obscure way. But not for checking whether a function has set its "out" parameters as documented.

For example, when calling `CSHA256::Finalize(result)` we don't set `result` to a dummy value before the call and check that it is not that dummy value after the call with `Assume()`. Same for any other function that has "out" parameters.

If you insist to check that `porphanTx` was set by the function, then I guess you should do the same with `from_peer`: `Assume(from_peer != -1)`.

michaelfolkson commented:
> NACK the addition of m_mutex_message_handling.

> Looks like neither one of us is convinced by the other's arguments. It's ok, it would have been too boring if everybody agreed on everything all the time.

@vasild: Are you in a position where you can ACK this PR overall (with reservations) or would you prefer not to ACK it? Either is fine (obviously), I just suspect other reviewers may be waiting to see if your concerns are addressed or not before spending time reviewing this.

ariard commented on src/net_processing.cpp:
```diff
@@ -2132,33 +2152,26 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
-void PeerManagerImpl::ProcessOrphanTx(std::set<uint256>& orphan_work_set)
+bool PeerManagerImpl::ProcessOrphanTx(Peer& peer)
 {
+    AssertLockHeld(m_mutex_message_handling);
```
At commit f86a525, I don't think this lock is useful to protect `vExtraTxnForCompact`, there is already one in `AddToCompactExtraTransactions` ? Though it might be useful at branch tip ?

ajtowns replied:
I was just adding `AssertLockHeld` to correspond with the `EXCLUSIVE_LOCKS_REQUIRED` annotation as a matter of course (see the recommendation in developer-notes.md).

ariard commented on src/net_processing.cpp:
```diff
@@ -3943,11 +3969,6 @@ bool PeerManagerImpl::ProcessMessages(CNode* pfrom, std::atomic<bool>& interrupt
 
-    {
-        LOCK(g_cs_orphans);
-        if (!peer->m_orphan_work_set.empty()) return true;
```
I think this introduce a lightweight behavior change.

Previously, if our peer has `fDisconnect=true` set up from a previous message processing and the orphan work set is still non-empty after the `ProcessOrphanTx` above, we would have return `false`.

From now on, if the orphan work set is still non-empty after `ProcessOrphanTx`, we're going to return `true`. 

Though i don't think it has any impact on allowing our peer to abuse more our CPU time, when we return from `ProcessMessages` to `ThreadMessageHandler`, we go to `SendMessages` which calls `MaybeDiscourageAndDisconnect` as first thing.

(Actually why the check on `fDisconnect` isn't first in `ProcessMessages`, should be before any kind of processing work attempted with `ProcessGetData`/`ProcessOrphanTx` ?)

ajtowns replied:
`fDisconnect` is already checked in net.cpp prior to `ProcessMessages` being called, but there's always a chance that there's a race and `fDisconnect` is set while in the middle of `ProcessMessages` for the peer.

ariard reviewed (commented):
Approach ACK, reviewed up to 1e3f227 so far

ariard commented on src/net.cpp:
```diff
@@ -2197,10 +2197,7 @@ void CConnman::ThreadMessageHandler()
-                LOCK(pnode->cs_sendProcessing);
-                m_msgproc->SendMessages(pnode);
-            }
+            m_msgproc->SendMessages(pnode);
```
Actually, sending block headers/compact blocks messages in parallel has been discussed a while back (iirc  https://bitcoincore.org/en/meetings/2018/05/03/, see "Call ProcessNewBlock asynchronously") though  we could insert some queued interface between a net_processing thread and multiple net threads ?

narula commented on src/net_processing.cpp:
```diff
@@ -4303,6 +4324,8 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
     if (!peer) return false;
     const Consensus::Params& consensusParams = m_chainparams.GetConsensus();
 
+    LOCK(m_mutex_message_handling);
```
>  /** Message handling mutex.
>    *  Message processing is single-threaded, so anything only accessed
>    *  by ProcessMessage(s) or SendMessages can be guarded by this mutex,
>    *  which guarantees it's only accessed by a single thread.

I'm confused. This mutex ends up protecting everything accessed in those functions. For example, in `SendMessages`, it also protects `MaybeDiscourageAndDisconnect` that accesses things like `peer.m_misbehavior_mutex`, which can be accessed by the validation thread. So this mutex is guarding more things than what is _only_ accessed by the `ProcessMessage(s)` and `SendMessages` thread.

I suppose it's safe that this mutex also ends up guarding things that are accessed by multiple threads, but I find the comment confusing.

ajtowns replied:
If you have:

```c++
Mutex mutex;
int x GUARDED_BY(mutex);
int y;
static void A() { LOCK(mutex); ++x; ++y; }
static void B() { LOCK(mutex); --x; }
void ThreadA() { for (int i = 0; i < 100; ++i) { A(); } }
void ThreadB() { for (int i = 0; i < 100; ++i) { B(); } }
```

then I'd say that only `x` is guarded by the mutex, even though `y` is only actually accessed while the mutex is held. If you later  introduce

```c++
static void C() { y *= 2; }
```

then that would still be safe/correct if `ThreadA` were to start calling `C()`, but `y` would no longer always be protected by the mutex; conversely if `ThreadB` were to start calling `C()`, then `y` would no longer be thread safe, but because it's not guarded by anything, the compiler would not catch that bug.

I guess I'd look at it more as mutexes protect **code** from being run simultaneously; it's the GUARDED_BY **annotations** that protect the variables. If the variables aren't annotated, they don't have any automatic protections, only manual protection by the programmer constantly being careful.

vasild replied:
> I guess I'd look at it more as mutexes protect __code__ from being run simultaneously

IMO mutexes protect variables, not code.

From https://en.wikipedia.org/wiki/Lock_(computer_science)
> ... a mechanism that enforces limits on access to a __resource__ ...
> ... each thread cooperates by acquiring the lock before accessing the __corresponding data__ ...

ajtowns replied:
Read your second quote carefully -- it's not the mutex/lock that protects the data, it's the threads that protect it by cooperatively acquiring the lock before accessing the data.

It's 99% true that the point of using a mutex is to protect data, but it's not the mutex itself that ensures the data is protected, it's how you use the mutex. It's like looking both ways before crossing the road -- if you don't do that, you'll probably walk into traffic; but looking alone isn't enough, you have to make sure you do it before crossing the road, and depending on what you see, delay crossing the road. The point of annotations here is to ensure at compile time that you're not metaphorically stepping into the road before looking, or literally not accessing variables that are only meant to be accessed by a particular thread from another thread.

(I think I've repeated this point enough now; if there's something not clear, feel free to ask, but I'm not seeing any point to going around in circles another time)

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -2396,6 +2407,14 @@ void PeerManagerImpl::ProcessBlock(CNode& pfrom, const std::shared_ptr<const CBl
 void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRecv,
```
do I understand correctly that after these changes `PeerManagerImpl::ProcessMessage` is a test-only function? If so, I think the comment in `net_processing.h` should be updated.

ajtowns replied:
`PeerManager::ProcessMessage` has always only been exposed for tests, and is already documented that way in the header file:

```c++
    /** Process a single message from a peer. Public for fuzz testing */
    virtual void ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRecv,
```

`PeerManagerImpl::ProcessMessage` is thus needed to complete the `PeerManager` interface.

vasild commented:
> > NACK the addition of m_mutex_message_handling.
> 
> > Looks like neither one of us is convinced by the other's arguments. It's ok, it would have been too boring if everybody agreed on everything all the time.
> 
> @vasild: Are you in a position where you can ACK this PR overall (with reservations) or would you prefer not to ACK it? Either is fine (obviously), I just suspect other reviewers may be waiting to see if your concerns are addressed or not before spending time reviewing this.

The rest of this PR looks good overall (modulo some minor other comments I left). But the mutex `m_mutex_message_handling` looks [really strange](https://github.com/bitcoin/bitcoin/pull/21527#discussion_r631895780) to me, with a [confusing comment](https://github.com/bitcoin/bitcoin/pull/21527#discussion_r654705472).

I would re-review carefully and probably ACK this PR if the mutex is removed. If needed later then it can be added in another PR.

jnewbery commented:
@ajtowns are you still working on this? I think this new net_processing lock is the best way we have to make progress with much of the net/net_processing/validation separation work.

ajtowns committed:
txorphanage: fix missing includes

ajtowns commented on src/net_processing.cpp:
```diff
@@ -3927,10 +3955,8 @@ bool PeerManagerImpl::ProcessMessages(CNode* pfrom, std::atomic<bool>& interrupt
+        if (ProcessOrphanTx(*peer)) return true;
     }
 
     if (pfrom->fDisconnect)
```
Changed to `return !pfrom->fDisconnect` to preserve the same behaviour.

ajtowns commented on src/txorphanage.cpp:
```diff
@@ -132,15 +134,19 @@ unsigned int TxOrphanage::LimitOrphans(unsigned int max_orphans)
+    LOCK(m_mutex);
+
+    // Get this peer's work set, ensuring it exists
+    std::set<uint256>& orphan_work_set = m_peer_work_set.emplace(peer, std::set<uint256>{}).first->second;
```
Changed to C++17's `try_emplace` which simplifies it slightly. Not really a fan of `map::operator[]`

ajtowns commented on src/txorphanage.h:
```diff
@@ -21,65 +21,76 @@ extern RecursiveMutex g_cs_orphans;
+    OrphanMap m_orphans GUARDED_BY(m_mutex);
+
+    /** Which peer provided a parent tx of orphans that need to be reconsidered */
+    std::map<NodeId, std::set<uint256>> m_peer_work_set GUARDED_BY(m_mutex);
```
Left as map; I don't think performance matters enough to justify using extra space for the hashmap; and longer term, using a multiindex rather than a bunch of different containers referencing each other is probably better.

ajtowns commented on src/net_processing.cpp:
```diff
@@ -329,7 +337,17 @@ class PeerManagerImpl final : public PeerManager
+     * @param[in,out]  peer             The peer whose orphan transactions we will reconsider. Generally only one
+     *                                  orphan will be reconsidered on each call of this function. This peer's set
+     *                                  may be added to if accepting an orphan causes its children to be
+     *                                  reconsidered.
```
Dropped the `out` which doesn't make sense. Knowing that this may increase the size of the set, not simply reduce it is valuable I think, so have left that text alone.

ajtowns commented on src/txorphanage.h:
```diff
@@ -21,65 +21,76 @@ extern RecursiveMutex g_cs_orphans;
 
     /** Erase all orphans included in or invalidated by a new block */
-    void EraseForBlock(const CBlock& block) LOCKS_EXCLUDED(::g_cs_orphans);
+    void EraseForBlock(const CBlock& block) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
```
Seems like something more for clang upstream if anything?

ajtowns commented:
Rebased past #21562. Minor updates per reviews.

glozow commented:
Concept ACK, I like the idea of txorphanage being internally thread-safe

ajtowns commented:
Rebased past #22976

ajtowns commented:
Rebased past a few conflicts.

ajtowns committed:
[move-only] net: move NetEventsInterface before CNode

This allows CNode members to be marked as guarded by the
g_mutex_msgproc_thread mutex.

ajtowns committed:
net: add thread safety annotations for CNode/Peer members accessed only via the msgproc thread

ajtowns committed:
net: mark TransportSerializer/m_serializer as const

The (V1)TransportSerializer instance CNode::m_serializer is used from
multiple threads via PushMessage without protection by a mutex. This
is only thread safe because the class does not have any mutable state,
so document that by marking the methods and the object as "const".

ajtowns committed:
net: note CNode members that are treated as const

m_permissionFlags and m_prefer_evict are treated as const -- they're
only set immediately after construction before any other thread has
access to the object, and not changed again afterwards. As such they
don't need to be marked atomic or guarded by a mutex; though it would
probably be better to actually mark them as const...

ajtowns commented:
Rebased with the `cs_sendProcessing` parts split out into #24474; marking as draft while that one gets reviewed.

asaneh reviewed (approved):
#22838 #24427 #21527 #24474 #22558

ajtowns commented:
Rebased to keep up with #24474

ajtowns committed:
net/net_processing: add missing thread safety annotations

ajtowns committed:
net: mark CNode unique_ptr members as const

Dereferencing a unique_ptr is not necessarily thread safe. The reason
these are safe is because their values are set at construction and do
not change later; so mark them as const and set them via the initializer
list to guarantee that.

ajtowns committed:
net_processing: add thread safety annotations for Peer members accessed only via the msgproc thread

ajtowns commented:
Closing this. If you'd like it resurrected please help https://github.com/bitcoin/bitcoin/pull/25174 make progress.

ajtowns commented:
Closing this. If you'd like it resurrected please help https://github.com/bitcoin/bitcoin/pull/25174 make progress.
